### PR TITLE
make php sdk use v3 and refactor tests

### PR DESCRIPTION
**Pull request by javaes**

### CODE REVIEW

File: figo/Config.php [hunk 1]
```diff
@@ -29,8 +29,8 @@
  */
 class Config {
 
-    /** @var string figo Connect server hostname */
-    public static $API_ENDPOINT = "api.figo.me";
+    /** @var string figo Connect server address. This should be the full base url of the API */
+    public static $API_ENDPOINT = "https://api.figo.me/v3";
```
Comment (ckoehn) [inline figo/Config.php]: Do we still support the legacy way of setting the `API_ENDPOINT` (just `api.figo.me`)? Do you handle trailing slashes?

File: figo/Connection.php [hunk 2]
```diff
@@ -107,7 +109,8 @@ public function query_api($path, array $data = null, $method='POST', $encode='ht
                         "Content-Type"   => $content_type,
                          "Content-Length" => strlen($data));
 
-        $request = new HttpsRequest($this->apiEndpoint, $this->fingerprints, $this->logger);
+        $request = new HttpsRequest($this->apiUrl['host'], $this->fingerprints, $this->logger);
+        $path = $this->apiUrl['path'] . $path;
```
Comment (ckoehn) [inline figo/Connection.php]: Can `$this->apiUrl['path']` be just `/ `? This might lead to errors then....

File: figo/Session.php [hunk 3]
```diff
@@ -93,8 +95,8 @@ public function query_api($path, array $data = null, $method = "GET") {
                          "Content-Type"   => "application/json",
                          "Content-Length" => strlen($data));
 
-        $request = new HttpsRequest($this->apiEndpoint, $this->fingerprints, $this->logger);
-
+        $request = new HttpsRequest($this->apiUrl['host'], $this->fingerprints, $this->logger);
+        $path = $this->apiUrl['path'] . $path;
```
Comment (ckoehn) [inline figo/Session.php]: Same comment as in the `Connection` class.

File: test/FigoTest.php [hunk 4]
```diff
@@ -29,210 +29,174 @@
...
+    protected static $session;
+
+    protected $access_token;
+    protected $account_id;
+    protected $sec_account_id;
+
+    public static function setUpBeforeClass()
+    {
+        self::$connection = new Connection(getenv("FIGO_CLIENT_ID"), getenv("FIGO_CLIENT_SECRET"),
+            "http://my-domain.org/redirect-url", self::$api_endpoint, self::$fingerprints);
```
Comment (ckoehn) [inline test/FigoTest.php]: Please use `http://example.com`

File: test/FigoTest.php [hunk 5]
```diff
@@ -29,210 +29,174 @@
 
 class SessionTest extends PHPUnit_Framework_TestCase {
 
-    protected $sut;
+    protected static $api_endpoint = "https://staging.figo.me/v3";
+    protected static $fingerprints = array("D0:03:9E:F0:8F:BD:48:67:86:71:CE:9D:A5:54:24:81:63:D7:D9:4D:ED:F1:6A:55:F0:52:C7:0A:AB:7B:B8:9D");
```
Comment (ckoehn) [inline test/FigoTest.php]: Shouldn't we make `$api_endpoint` and `$fingerprints` configurable too?

File: test/FigoTest.php [hunk 6]
```diff
@@ -29,210 +29,174 @@
...
+        $response = self::$connection->native_client_login(self::$email, self::$password);
+        $access_token = $response["access_token"];
+        self::$session = new Session($access_token, self::$api_endpoint, self::$fingerprints);
     }
 
-    public function test_accounts() {
-        $accounts = $this->sut->get_accounts();
-        $this->assertGreaterThan(0, count($accounts));
+    public static function tearDownAfterClass()
+    {
```
Comment (ckoehn) [inline test/FigoTest.php]: Please unify the way you use curly braces. Currently we have different styles here:

```php
public static function tearDownAfterClass() {
    ...
```
vs
```php
public static function tearDownAfterClass() 
{
    ...
```

File: test/FigoTest.php [hunk 7]
```diff
@@ -29,210 +29,174 @@
...
+
+        $accounts = $this::$session->get_accounts();
+        $this->assertEquals([], $accounts);
+    }
+
+    public function test_setup_account()    {
+        $response = $this::$session->setup_bank_account("90090042", array("figo", "figo"), array());
+        for($i = 0; $i <= 20; $i++)   {
+            $task_state = $this::$session->get_task_state($response['task_token']);
+            if($task_state['is_ended'] == true) {
```
Comment (ckoehn) [inline test/FigoTest.php]: What if the task `is_erroneous`? Shouldn't we add an assert for this?

File: test/FigoTest.php [hunk 8]
```diff
@@ -29,210 +29,174 @@
...
 
     public function test_error_handling() {
          $this->setExpectedException('Exception');
-         $this->sut->get_sync_url("http://localhost:3003/", "");
+         $this::$session->get_sync_url("random", "");
     }
 
     public function test_sync_url() {
-        $sync_url = $this->sut->get_sync_url("qwe", "qew");
+        $sync_url = $this::$session->get_sync_url("http://localhost:3003/", "qew");
```
Comment (ckoehn) [inline test/FigoTest.php]: `example.com`?!

File: figo/Config.php (hunk 1 above)
Comment (javaes) [reply figo/Config.php]: yes we support both ways now

File: figo/Connection.php (hunk 2 above)
Comment (javaes) [reply figo/Connection.php]: not possible anymore because of `parse_api_endpoint()`

File: test/FigoTest.php (hunk 7 above)
Comment (javaes) [reply test/FigoTest.php]: assertion added

File: figo/Config.php [hunk 9]
```diff
@@ -42,7 +42,7 @@ class Config {
     /**
      * @var string Version of this SDK, used in user agent for API requests
      */
-    public static $SDK_VERSION = '1.3.0';
+    public static $SDK_VERSION = '1.4.0';
```
Comment (ckoehn) [inline figo/Config.php]: Please use `2.0.0`.

File: figo/utils.php [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+    $api_endpoint = rtrim($api_endpoint, "/");
+    if (substr($api_endpoint, 0, 8) != "https://")    {
+        $api_endpoint = "https://" . $api_endpoint;
+    }
+    $api_url = parse_url($api_endpoint);
+    if (array_key_exists("path", $api_url) == false)    {
+        $api_url["path"] = "";
+    }
+    return $api_url;
+}
```
Comment (ckoehn) [inline figo/utils.php]: I think you need to close the `php` tag. I'm not sure how picky php is...

File: composer.json [hunk 11]
```diff
@@ -2,7 +2,7 @@
   "name": "figo/figo",
   "description": "API wrapper for figo Connect.",
   "homepage": "https://github.com/figo-connect/php-figo",
-  "version": "1.3.0",
+  "version": "1.4.0",
```
Comment (ckoehn) [inline composer.json]: `2.0.0`

File: figo/utils.php (hunk 10 above)
Comment (javaes) [reply figo/utils.php]: according to this and my ide it seems to be a good practice to omit it: https://stackoverflow.com/questions/3219383/why-do-some-scripts-omit-the-closing-php-tag

File: figo/Config.php [hunk 12]
```diff
@@ -42,7 +42,7 @@ class Config {
     /**
      * @var string Version of this SDK, used in user agent for API requests
      */
-    public static $SDK_VERSION = '1.3.0';
+    public static $SDK_VERSION = '2.0.0';
```
Comment (PATROMO) [inline figo/Config.php]: Version should be 3.0.0!

Comment (berend) [reply figo/Config.php]: You are right! Thanks for pointing that out.

Comment (berend) [reply figo/Config.php]: just released 3.0.1. And made a note to automate this, to have only one place to store the version string...

Comment (PATROMO) [reply figo/Config.php]: 👍

**Pull request merged**